File: src/lib/utils/virtualTags.ts
/**
 * Virtual Tag System
 *
 * Virtual tags are special tags with prefixes that have semantic meaning.
 * They are stored with their prefix but displayed differently in the UI.
 *
 * Current virtual tags:
 * - t: (title) - Custom note title
 *
 * Future virtual tags could include:
 * - d: (due date)
 * - p: (priority)
 * - etc.
 */

export interface VirtualTagConfig {
  /** The prefix identifying this virtual tag type (e.g., 't:') */
  prefix: string;
  /** The i18n key for the display name (e.g., 'tags.virtual.title') */
  displayKey: string;
  /** Whether only one tag of this type is allowed per note */
  singular: boolean;
  /** Visual styling for the tag */
  style: 'italic' | 'bold';
}

/**
 * Configuration for all virtual tag types.
 * Keys are the virtual tag type names (used internally).
 */
export const VIRTUAL_TAGS: Record<string, VirtualTagConfig> = {
  title: {
    prefix: 't:',
    displayKey: 'tags.virtual.title',
    singular: true,
    style: 'italic',
  },
  // Future virtual tags can be added here
};

/** The prefix for title tags */
export const TITLE_TAG_PREFIX = 't:';

/**
 * Check if a tag is a virtual tag (has a known prefix).
 * @param tag The tag to check
 * @returns true if the tag is a virtual tag
 */
export function isVirtualTag(tag: string): boolean {
  return Object.values(VIRTUAL_TAGS).some((config) => tag.startsWith(config.prefix));
}

/**
 * Get the virtual tag config for a tag, if any.
 * @param tag The tag to check
 * @returns The config if tag is virtual, undefined otherwise
 */
export function getVirtualTagConfig(tag: string): VirtualTagConfig | undefined {
  return Object.values(VIRTUAL_TAGS).find((config) => tag.startsWith(config.prefix));
}

/**
 * Get the display information for a virtual tag.
 * @param tag The virtual tag
 * @returns Display info (displayKey, value, style) or null if not a virtual tag
 */
export function getVirtualTagDisplay(
  tag: string
): { displayKey: string; value: string; style: string } | null {
  const config = getVirtualTagConfig(tag);
  if (!config) return null;
  return {
    displayKey: config.displayKey,
    value: tag.substring(config.prefix.length),
    style: config.style,
  };
}

/**
 * Filter tags to get only regular (non-virtual) tags.
 * @param tags Array of tags
 * @returns Array of tags that are not virtual tags
 */
export function getRegularTags(tags: string[]): string[] {
  return tags.filter((tag) => !isVirtualTag(tag));
}

/**
 * Filter tags to get only virtual tags.
 * @param tags Array of tags
 * @returns Array of virtual tags
 */
export function getVirtualTags(tags: string[]): string[] {
  return tags.filter((tag) => isVirtualTag(tag));
}

/**
 * Get the virtual tag type name for a tag.
 * @param tag The tag to check
 * @returns The type name (e.g., 'title') or undefined if not virtual
 */
export function getVirtualTagType(tag: string): string | undefined {
  for (const [type, config] of Object.entries(VIRTUAL_TAGS)) {
    if (tag.startsWith(config.prefix)) {
      return type;
    }
  }
  return undefined;
}

File: tui/src/models/virtual_tags.rs
//! Virtual Tag System
//!
//! Virtual tags are special tags with prefixes that have semantic meaning.
//! They are stored with their prefix but displayed differently in the UI.
//!
//! Current virtual tags:
//! - t: (title) - Custom note title
//!
//! Future virtual tags could include:
//! - d: (due date)
//! - p: (priority)
//! - etc.

use rust_i18n::t;

/// The prefix for title tags
pub const TITLE_TAG_PREFIX: &str = "t:";

/// Check if a tag is a virtual tag
pub fn is_virtual_tag(tag: &str) -> bool {
    tag.starts_with(TITLE_TAG_PREFIX)
}

/// Get regular (non-virtual) tags
pub fn get_regular_tags(tags: &[String]) -> Vec<&String> {
    tags.iter().filter(|t| !is_virtual_tag(t)).collect()
}

/// Get all virtual tags
pub fn get_virtual_tags(tags: &[String]) -> Vec<&String> {
    tags.iter().filter(|t| is_virtual_tag(t)).collect()
}

/// Extract title from note, checking for t: tag first, then first non-empty line
pub fn get_note_title(content: &str, tags: &[String]) -> String {
    // Check for title tag first
    if let Some(title_tag) = tags.iter().find(|t| t.starts_with(TITLE_TAG_PREFIX)) {
        let title = title_tag.strip_prefix(TITLE_TAG_PREFIX).unwrap_or("").trim();
        if !title.is_empty() {
            return title.to_string();
        }
    }

    // Fall back to first non-empty line
    content
        .lines()
        .find(|line| !line.trim().is_empty())
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|| t!("note.untitled").to_string())
}

/// Check if note has custom title tag
pub fn has_custom_title(tags: &[String]) -> bool {
    tags.iter().any(|t| t.starts_with(TITLE_TAG_PREFIX))
}

/// Get the title tag value (without prefix), or None
pub fn get_title_tag_value(tags: &[String]) -> Option<&str> {
    tags.iter()
        .find(|t| t.starts_with(TITLE_TAG_PREFIX))
        .map(|t| t.strip_prefix(TITLE_TAG_PREFIX).unwrap_or(""))
}

/// Set title tag, removing existing. Returns new tags.
pub fn set_title_tag(tags: &[String], title: &str) -> Vec<String> {
    let mut filtered: Vec<String> = tags
        .iter()
        .filter(|t| !t.starts_with(TITLE_TAG_PREFIX))
        .cloned()
        .collect();

    if !title.trim().is_empty() {
        filtered.push(format!("{}{}", TITLE_TAG_PREFIX, title.trim()));
    }

    filtered
}

/// Remove title tag. Returns new tags.
pub fn remove_title_tag(tags: &[String]) -> Vec<String> {
    tags.iter()
        .filter(|t| !t.starts_with(TITLE_TAG_PREFIX))
        .cloned()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_virtual_tag() {
        assert!(is_virtual_tag("t:My Title"));
        assert!(is_virtual_tag("t:"));
        assert!(!is_virtual_tag("project"));
        assert!(!is_virtual_tag("title"));
        assert!(!is_virtual_tag("t"));
    }

    #[test]
    fn test_get_regular_tags() {
        let tags = vec![
            "project".to_string(),
            "t:My Title".to_string(),
            "work".to_string(),
        ];
        let regular = get_regular_tags(&tags);
        assert_eq!(regular.len(), 2);
        assert_eq!(regular[0], "project");
        assert_eq!(regular[1], "work");
    }

    #[test]
    fn test_get_note_title_with_tag() {
        let content = "First line content";
        let tags = vec!["t:Custom Title".to_string(), "work".to_string()];
        assert_eq!(get_note_title(content, &tags), "Custom Title");
    }

    #[test]
    fn test_get_note_title_without_tag() {
        let content = "First line content\nSecond line";
        let tags = vec!["work".to_string()];
        assert_eq!(get_note_title(content, &tags), "First line content");
    }

    #[test]
    fn test_get_note_title_empty_tag() {
        let content = "First line content";
        let tags = vec!["t:".to_string()];
        assert_eq!(get_note_title(content, &tags), "First line content");
    }

    #[test]
    fn test_has_custom_title() {
        let tags_with = vec!["t:Title".to_string(), "work".to_string()];
        let tags_without = vec!["work".to_string()];
        assert!(has_custom_title(&tags_with));
        assert!(!has_custom_title(&tags_without));
    }

    #[test]
    fn test_set_title_tag() {
        let tags = vec!["project".to_string(), "t:Old".to_string()];
        let new_tags = set_title_tag(&tags, "New Title");
        assert_eq!(new_tags.len(), 2);
        assert!(new_tags.contains(&"project".to_string()));
        assert!(new_tags.contains(&"t:New Title".to_string()));
    }

    #[test]
    fn test_set_title_tag_empty() {
        let tags = vec!["project".to_string(), "t:Old".to_string()];
        let new_tags = set_title_tag(&tags, "");
        assert_eq!(new_tags.len(), 1);
        assert_eq!(new_tags[0], "project");
    }

    #[test]
    fn test_remove_title_tag() {
        let tags = vec!["project".to_string(), "t:Title".to_string()];
        let new_tags = remove_title_tag(&tags);
        assert_eq!(new_tags.len(), 1);
        assert_eq!(new_tags[0], "project");
    }
}
